lib/external_count_fetcher.rb

```diff
@@ -6,61 +6,79 @@
 #   # => {}
 #
 #   ExternalCountFecther.new.counts
-#   # => {supporter_count: 12, volunteer_count: 12, called_voters_count: 12, called_reps_count: 12, house_supporters: 12, senate_supporters: 12, donations_total: 12, donations_count: 12}
+#   # => {supporter_count: 12, volunteer_count: 12, called_voters_count: 12, reps_calls_count: 12, house_supporters: 12, senate_supporters: 12, donations_total: 12, donations_count: 12}
  class ExternalCountFetcher
-  attr_accessor :counter_hash
+  include Redis::Objects
 
-  COUNTER_HASH_KEYS = [:supporter_count, :volunteer_count, :called_voters_count, :called_reps_count, :house_supporters, :senate_supporters, :donations_total, :donations_count]
+  REDIS_KEYS = [:supporter_count, :volunteer_count, :called_voters_count, :reps_calls_count, :house_supporters, :senate_supporters, :donations_total, :donations_count]
+  REDIS_EXPIRE_SECONDS = 3.hours.to_i
 
+  REDIS_KEYS.each do |key|
+    counter key, :expiration => REDIS_EXPIRE_SECONDS
+  end
 
-  def initialize
-    self.counter_hash = Redis::HashKey.new('external_counts', :expiration => 3.hours) #Note: expiration is only set when hash becomes non-empty
+  def counts!
+    fetch_empty_counts!
+    counts
   end
 
   # Public: fetches all counts from external services. This is a high latency call and shouldn't only be called from a user request as a last resort
-  # MARIO TODO: break out into multiple threads and redis keys, handle error cases
-  def fetch_all!
-    fetch_donations_total
-    fetch_donations_count
-    fetch_supporter_counts
-    counter_hash.all
+  #
+  # reset - true|false  - if true, will fetch counters even if already set.
+  def fetch_empty_counts!(reset: false)
+    threads = []
+    REDIS_KEYS.each do |key|
+      threads << Thread.new do
+        fetch_count(counter_key: key, reset: reset) if !redis_counter(key).exists? || reset
+      end
+    end
+    threads.each &:join
   end
 
-  # Public: fetches all counts from external services. If the redis cache of counts is empty it will call `fetch_all!`
+  # Public: fetches all counts from external services.
   def counts
-    counter_hash.empty? ? fetch_all! : counter_hash.all
+    REDIS_KEYS.each_with_object({}) do |key, hash|
+      hash[key] = redis_counter(key).value
+    end
   end
 
   private
 
-  def fetch_donations_total
-    self.counter_hash[:donations_count] = Integration::PledgeService.donations_count
-  end
-
-  def fetch_donations_count
-    self.counter_hash[:donations_total] = Integration::PledgeService.donations_total
-  end
-
-  def fetch_supporter_counts
-    Integration::NationBuilder.list_counts.each do |key,count|
-      self.counter_hash[key.to_sym] = count
+  def fetch_count(counter_key:, reset: false)
+    if counter_key =~ /supporter_count|volunteer_count/
+      fetch_supporter_counts
+    else
+      count = case counter_key
+        when :donations_count     then Integration::PledgeService.donations_count
```

Comment:
Line is too long. [81/80]<br>Indent `when` as deep as `case`.<br>Unnecessary spacing detected.